### ISSUE TITLE
Bug : URL redirection does not seem to work properly in UDC

### ISSUE THREAD

**Issue by sanchit-saini** [ucscGenomeBrowser/kent#53]
Hello,

[rtracklayer](https://github.com/lawremi/rtracklayer) relies on the kent library and It seems UDC(URL data cache) does not handle URL redirection properly that is causing https://github.com/lawremi/rtracklayer/issues/42 issue.

I tested [udcFileMayOpen](https://github.com/ucscGenomeBrowser/kent/blob/b8032901670f0266e9c228bd9fda6a74463c65b9/src/lib/udc.c#L1135) function with "http://bedbase.org/api/bed/78c0e4753d04b238fc07e4ebe5a02984/file/bigbedfile"
For ease of testing, I inserted the following code segment in the `src/utils/bedToBigBed` utility (in the usage function).
```C
void usage()
/* Explain usage and exit. */
{
struct udcFile *udcTestFile =
udcFileMayOpen("http://bedbase.org/api/bed/78c0e4753d04b238fc07e4ebe5a02984/file/bigbedfile", udcDefaultDir());
if (udcTestFile == NULL)
    printf("Not Working\n");
else
    printf("Working\n");
...
```
After compilation received `Not working` as output. This function [udcInfoViaHttp](https://github.com/ucscGenomeBrowser/kent/blob/b9b84244ae9fbfa6d81c92028634be26e8a5898f/src/lib/udc.c#L492-#L642)  seems related to the URL direction. I would be happy to help.

Thanks!

**Comment by genome-www** [comment]
Hi there,

The issue here is the server is not responding properly to HEAD
requests, as evidenced below:

curl -I http://bedbase.org/api/bed/78c0e4753d04b238fc07e4ebe5a02984/file/bigbedfile

HTTP/1.1 404 Not Found
Server: nginx/1.19.6
Date: Tue, 09 Mar 2021 15:48:17 GMT
Content-Type: text/plain; charset=utf-8
Content-Length: 9
Connection: keep-alive


If the server responds with the proper redirect HTTP status code then
this should work fine.

Please let us know if you have any further questions.

Christopher Lee
UCSC Genomics Institute

**Comment by nsheff** [comment]
Just to add for future reference, this is because bedbase is built using fastAPI, which does not provide HEAD responses automatically for GET endpoints. The issue is documented here: https://github.com/tiangolo/fastapi/issues/1773

In practice you don't always need the HEAD response to follow the URL redirect;  browsers, wget for example will still work, it just depends on whether the requester relies on the HEAD response or doesn't -- but the UCSC code does require it. Anyway, we will solve it on our end for now and and per that issue, fastAPI should eventually support HEAD automatically.

**Comment by genome-www** [comment]
Yes, fastapi was mostly built for creating fast web APIs. HEAD requests are
not important in this context.
